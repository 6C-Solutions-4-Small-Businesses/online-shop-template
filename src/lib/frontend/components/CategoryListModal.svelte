<script lang="ts">
    import type {SvelteComponent} from 'svelte'
    import CloseIcon from '$lib/frontend/components/icons/CloseIcon.svelte'
    import type {CategorySummaryPresentation} from '$lib/frontend/presentations/CategorySummaryPresentation'
    import {writable, type Writable} from 'svelte/store'
    import {goto} from '$app/navigation'

    export let parent: SvelteComponent
    export let categories: CategorySummaryPresentation []

    const indexedCategories: Writable<Record<string, {
        id: string,
        name: string
    }[]>> = writable(indexCategories(categories))

    function indexCategories(categories: CategorySummaryPresentation[]): Record<string, {
        id: string,
        name: string
    }[]> {
        const categorized: Record<string, {
            id: string,
            name: string
        }[]> = {}

        categories.forEach(category => {
            const firstLetter = category.name[0].toUpperCase()
            if (!categorized[firstLetter]) {
                categorized[firstLetter] = []
            }
            categorized[firstLetter].push({
                id: category.id,
                name: category.name
            })
        })

        for (const key in categorized) {
            categorized[key].sort()
        }

        const sortedKeys = Object.keys(categorized).sort()
        return sortedKeys.reduce((acc: Record<string, {
            id: string,
            name: string
        }[]>, key: string) => {
            acc[key] = categorized[key]
            return acc
        }, {})
    }

    async function goToProductsPage(categoryId: string, categoryName: string): Promise<void> {
        parent.onClose()
        await goto(`/category/${categoryId}`, {
            state: {
                name: categoryName
            },
        })
    }
</script>

<div class="flex flex-col gap-2.5 w-11/12 sm:w-11/12 lg:w-10/12 py-6 px-7 bg-white rounded-4xl">
    <header class="flex items-center justify-end">
        <button data-testid="category-list-modal-close-button" on:click={parent.onClose}>
            <CloseIcon/>
        </button>
    </header>
    <div class="px-6 gap-2.5">
        <div class="flex justify-start">
            <span class="text-primary uppercase text-2xl">Categories</span>
        </div>
        <div class="h-[654px] py-9 px-5 grid gap-4 sm:grid-cols-2 md:grid-cols-3 lg:grid-cols-4 2xl:grid-cols-5 3xl:grid-cols-6 overflow-y-scroll">
            {#each Object.keys($indexedCategories) as startingLetter}
                <div class="min-h-[229px]">
                    <span class="text-black font-bold text-[50px] leading-12">{startingLetter}</span>
                    <ul role="list"
                        class="list-disc marker:text-primary text-black font-medium text-2xl list-outside ml-4">
                        {#each $indexedCategories[startingLetter] as category}
                            <li class="h-10">
                                <button class="h-full text-left flex items-center"
                                        on:click={() => goToProductsPage(category.id, category.name)}>
                                    <span class="line-clamp-1">{category.name}</span>
                                </button>
                            </li>
                        {/each}
                    </ul>
                </div>
            {/each}
        </div>
    </div>
</div>