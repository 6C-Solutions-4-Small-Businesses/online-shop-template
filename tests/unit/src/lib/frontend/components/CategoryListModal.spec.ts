import {describe, expect, vi} from "vitest";
import {fireEvent, render} from "@testing-library/svelte";
import CategoryListModal from "$lib/frontend/components/CategoryListModal.svelte";
import {mock, type MockProxy} from "vitest-mock-extended";
import type {SvelteComponent} from "svelte";

describe('Category list modal', () => {
    it('should have match snapshot', () => {
        const {container} = render(CategoryListModal, {
            categories: [
                {id: 'd223bf35-9418-445d-821f-63d260991d3e', name: 'Accessoire', type: "Common"},
                {id: '1482c135-b8e6-4b6a-b125-e8c2b5e9ca7c', name: 'Biscuit', type: "Common"},
                {id: 'baa0dfe4-8d22-4611-b268-62143e96e65f', name: 'Boisson Gaseuze', type: "Common"},
                {id: '84a4a6ff-b5d3-4f19-9368-b5e438d58207', name: 'Collation', type: "Common"},
                {id: '9ff12202-4488-4854-a320-8e4899f96a69', name: 'Condiment', type: "Common"},
            ],
            parent: mock<SvelteComponent>()
        })

        expect(container).toMatchSnapshot()
    })

    it('should have called parent.onClose when the close button is clicked', async () => {
        const parentMock: MockProxy<SvelteComponent> = mock<SvelteComponent>({
            onClose: vi.fn()
        })
        const {getByTestId} = render(CategoryListModal, {
            categories: [
                {id: 'd223bf35-9418-445d-821f-63d260991d3e', name: 'Accessoire', type: "Common"},
                {id: '1482c135-b8e6-4b6a-b125-e8c2b5e9ca7c', name: 'Biscuit', type: "Common"},
                {id: 'baa0dfe4-8d22-4611-b268-62143e96e65f', name: 'Boisson Gaseuze', type: "Common"},
                {id: '84a4a6ff-b5d3-4f19-9368-b5e438d58207', name: 'Collation', type: "Common"},
                {id: '9ff12202-4488-4854-a320-8e4899f96a69', name: 'Condiment', type: "Common"},
            ],
            parent: parentMock
        })
        const closeButton = getByTestId('category-list-modal-close-button');

        await fireEvent.click(closeButton);

        expect(parentMock.onClose).toHaveBeenCalled();
    });

    it('should have called parent.onClose when navigating to products page', async () => {
        const parentMock: MockProxy<SvelteComponent> = mock<SvelteComponent>({
            onClose: vi.fn()
        })
        const {getByText} = render(CategoryListModal, {
            categories: [
                {id: 'd223bf35-9418-445d-821f-63d260991d3e', name: 'Accessoire', type: "Common"},
                {id: '1482c135-b8e6-4b6a-b125-e8c2b5e9ca7c', name: 'Biscuit', type: "Common"},
                {id: 'baa0dfe4-8d22-4611-b268-62143e96e65f', name: 'Boisson Gaseuze', type: "Common"},
                {id: '84a4a6ff-b5d3-4f19-9368-b5e438d58207', name: 'Collation', type: "Common"},
                {id: '9ff12202-4488-4854-a320-8e4899f96a69', name: 'Condiment', type: "Common"},
            ],
            parent: parentMock
        })
        const closeButton = getByText('Boisson Gaseuze');

        await fireEvent.click(closeButton);

        expect(parentMock.onClose).toHaveBeenCalled();
    });
})

vi.mock('$app/navigation', () => ({
    goto: vi.fn(),
}));
