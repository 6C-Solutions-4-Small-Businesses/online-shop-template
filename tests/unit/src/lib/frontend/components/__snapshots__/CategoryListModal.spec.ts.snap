// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Category list modal > should have match snapshot 1`] = `
<div>
  <div
    class="flex flex-col gap-2.5 w-11/12 sm:w-11/12 lg:w-10/12 py-6 px-7 bg-white rounded-4xl"
  >
    <header
      class="flex items-center justify-end"
    >
      <button
        data-testid="category-list-modal-close-button"
      >
        <div>
          <svg
            fill="none"
            height="14"
            viewBox="0 0 20 20"
            width="14"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M1 19L19 1M1 1L19 19"
              stroke="#0A3B2F"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="1.5"
            />
          </svg>
        </div>
      </button>
    </header>
     
    <div
      class="px-6 gap-2.5"
    >
      <div
        class="flex justify-start"
      >
        <span
          class="text-primary uppercase text-2xl"
        >
          Categories
        </span>
      </div>
       
      <div
        class="h-[654px] py-9 px-5 grid gap-4 sm:grid-cols-2 md:grid-cols-3 lg:grid-cols-4 2xl:grid-cols-5 3xl:grid-cols-6 overflow-y-scroll"
      >
        <div
          class="min-h-[229px]"
        >
          <span
            class="text-black font-bold text-[50px] leading-12"
          >
            A
          </span>
           
          <ul
            class="list-disc marker:text-primary text-black font-medium text-2xl list-outside ml-4"
            role="list"
          >
            <li
              class="h-10"
            >
              <button
                class="h-full text-left flex items-center"
              >
                <span
                  class="line-clamp-1"
                >
                  Accessoire
                </span>
              </button>
               
            </li>
          </ul>
           
        </div>
        <div
          class="min-h-[229px]"
        >
          <span
            class="text-black font-bold text-[50px] leading-12"
          >
            B
          </span>
           
          <ul
            class="list-disc marker:text-primary text-black font-medium text-2xl list-outside ml-4"
            role="list"
          >
            <li
              class="h-10"
            >
              <button
                class="h-full text-left flex items-center"
              >
                <span
                  class="line-clamp-1"
                >
                  Biscuit
                </span>
              </button>
               
            </li>
            <li
              class="h-10"
            >
              <button
                class="h-full text-left flex items-center"
              >
                <span
                  class="line-clamp-1"
                >
                  Boisson Gaseuze
                </span>
              </button>
               
            </li>
          </ul>
           
        </div>
        <div
          class="min-h-[229px]"
        >
          <span
            class="text-black font-bold text-[50px] leading-12"
          >
            C
          </span>
           
          <ul
            class="list-disc marker:text-primary text-black font-medium text-2xl list-outside ml-4"
            role="list"
          >
            <li
              class="h-10"
            >
              <button
                class="h-full text-left flex items-center"
              >
                <span
                  class="line-clamp-1"
                >
                  Collation
                </span>
              </button>
               
            </li>
            <li
              class="h-10"
            >
              <button
                class="h-full text-left flex items-center"
              >
                <span
                  class="line-clamp-1"
                >
                  Condiment
                </span>
              </button>
               
            </li>
          </ul>
           
        </div>
      </div>
    </div>
  </div>
</div>
`;
